### PR TITLE
DC-457, DC-459: Add export to workspace

### PR DESCRIPTION
**Pull request by fboulnois**

### CODE REVIEW

File: service/src/main/java/bio/terra/catalog/rawls/RawlsService.java
Comment (fboulnois) [inline service/src/main/java/bio/terra/catalog/rawls/RawlsService.java:144]: What should we do if we get an error? There are many errors that are possible, but one obvious one is what happens if you export to workspace with entities that already exist in the destination workspace?

**Review by s-rubenstein — COMMENTED** [review]
Do we have a ticket for integration tests? That should probably be done in a follow up PR.

File: service/src/main/java/bio/terra/catalog/service/DatasetService.java
Comment (fboulnois) [reply service/src/main/java/bio/terra/catalog/service/DatasetService.java:259]: I'm not totally sure about this one, but I implemented the others.

File: service/src/main/java/bio/terra/catalog/api/DatasetApiController.java
Comment (pshapiro4broad) [inline service/src/main/java/bio/terra/catalog/api/DatasetApiController.java:94]: Is there a reason to convert the workspace ID into a string at this layer? Would it be incorrect to make the `DatasetService` API use UUIDs? I think it would be nice to use a more specific type in our code if possible.

File: service/src/main/java/bio/terra/catalog/rawls/RawlsService.java
Comment (pshapiro4broad) [inline service/src/main/java/bio/terra/catalog/rawls/RawlsService.java:120]: Would it be worth making a static method that converted from `WorkspaceDetails` to `WorkspaceName` to avoid the code duplication?

File: service/src/main/java/bio/terra/catalog/rawls/RawlsService.java
Comment (pshapiro4broad) [inline service/src/main/java/bio/terra/catalog/rawls/RawlsService.java:137]: This ignores `EntityCopyResponse` which may contain conflicts information, I assume generated when exporting into a workspace that already contains some entities. I don't think we should hold up this work to handle conflicts but maybe we should have a ticket to look at adding this support in the future?

File: service/src/main/java/bio/terra/catalog/service/DatasetService.java
Comment (pshapiro4broad) [reply service/src/main/java/bio/terra/catalog/service/DatasetService.java:259]: I agree that we shouldn't do a permission check here. At best it's duplication of effort, where we're doing the same check as the service; at worst we might prevent an operation that the service would allow.

File: service/src/test/java/bio/terra/catalog/api/DatasetApiControllerTest.java
Comment (pshapiro4broad) [inline service/src/test/java/bio/terra/catalog/api/DatasetApiControllerTest.java:199]: I assume this has the same effect as doing a `verify()` after the `perform()`. Is there a reason to prefer one form over the other? It looks like other tests in this class use `verify()` to do this.

File: service/src/test/java/bio/terra/catalog/datarepo/DatarepoServiceTest.java
Comment (pshapiro4broad) [inline service/src/test/java/bio/terra/catalog/datarepo/DatarepoServiceTest.java:171]: This is fine although it's only needed because this PR adds code to `DatasetService`. Alternatively you could make TDR snapshots the same as external (no-op in the switch) for now.

File: service/src/main/java/bio/terra/catalog/service/DatasetService.java
Comment (s-rubenstein) [reply service/src/main/java/bio/terra/catalog/service/DatasetService.java:259]: I think the general rule we've been following is to not check permissions if the underlying storage system is going to do it for us - it duplicates the check -> https://broadworkbench.atlassian.net/browse/DC-315 is the ticket that describes the why/history of that decision.

File: service/src/main/java/bio/terra/catalog/api/DatasetApiController.java
Comment (s-rubenstein) [inline service/src/main/java/bio/terra/catalog/api/DatasetApiController.java:93]: Do you think it would be worth filing a tech debt ticket to change the other param names to be `datasetId` as well?

File: service/src/main/java/bio/terra/catalog/datarepo/DatarepoService.java
Comment (pshapiro4broad) [inline service/src/main/java/bio/terra/catalog/datarepo/DatarepoService.java:152]: I think `exportSnapshot()` is a better name for this method as the operation isn't catalog dataset specific within the service code layer.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (pshapiro4broad) [inline service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:69]: I think in practice external source IDs will be UUIDs minted either within catalog or by catalog's callers. But for this test it shouldn't matter what the source ID is. Is there a reason why `DatasetServiceTest.sourceId` wouldn't work?

File: service/src/main/java/bio/terra/catalog/api/DatasetApiController.java
Comment (fboulnois) [reply service/src/main/java/bio/terra/catalog/api/DatasetApiController.java:93]: Done.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (pshapiro4broad) [inline service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:83]: Not sure why these fields are static.. do they rely on data which changes in each test run? (looking at the change history, it's probably my fault that they weren't static before)

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (pshapiro4broad) [inline service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:301]: I don't think this is necessary, as I'd expect `@Mock` objects to be re-initialized on every test run. Does it work without `reset()`? Looking at the code, this is also done in `testGetMetadataUsingTdrPermission()` but I think that use is also unnecessary and should've been removed when `mockDataset()` was added.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (pshapiro4broad) [inline service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:314]: This test doesn't appear to be specific to external datasets. It's verifying that `exportDataset()` is doing a permission check, which fails because the current user isn't an admin user. If you mock that in sam, then you can write a test that verifies that trying to export an external dataset is a no-op.
```
    when(samService.hasGlobalAction(user, SamAction.READ_ANY_METADATA)).thenReturn(true);
```

But I wouldn't recommend doing that as we haven't worked on external datasets yet so there's not much value to us to verify our APIs behavior when using external datasets.

So you could keep this test but make it use either the TDR or workspace dataset.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (pshapiro4broad) [inline service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:334]: I don't think this is necessary with strict subbing. If it's not called, you should get a test failure because of the mock on line 332.

**Review by pshapiro4broad — APPROVED** [review]
Many comments, looks good overall.

File: service/src/main/java/bio/terra/catalog/rawls/RawlsService.java
Comment (fboulnois) [reply service/src/main/java/bio/terra/catalog/rawls/RawlsService.java:137]: I added a ticket for this.

File: service/src/main/java/bio/terra/catalog/service/DatasetService.java
Comment (fboulnois) [reply service/src/main/java/bio/terra/catalog/service/DatasetService.java:259]: Okay, I've removed this.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (fboulnois) [reply service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:69]: Since I created a mock external dataset, I also created a mock `externalSourceId`. The regular `sourceId` could work as long as the mock external dataset also contained that `sourceId`. It would be confusing to me if I was looking in the debugger and I saw a mock external dataset with just the  regular `sourceId`, though (in fact this happened to me initially).

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (fboulnois) [reply service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:301]: It does work without reset, I've removed those.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (fboulnois) [reply service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java:83]: Do you mean "not sure why these fields **aren't** static"? If so, should I make them all static?

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (fboulnois) [reply service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java]: I ended up removing this due to removing the test for external datasets.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (fboulnois) [reply service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java]: I removed this test.

File: service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java
Comment (pshapiro4broad) [reply service/src/test/java/bio/terra/catalog/service/DatasetServiceTest.java]: > Do you mean "not sure why these fields aren't static"? If so, should I make them all static?

Yes, that's what I meant. If you could make them all static I think that would be better.

**Comment by pshapiro4broad** [comment]
Before merging could you look at the two "code smells" reported by sonar? https://sonarcloud.io/project/issues?id=terra-data-catalog&pullRequest=114&resolved=false&types=CODE_SMELL

File: service/src/main/java/bio/terra/catalog/rawls/RawlsService.java
Comment (s-rubenstein) [inline service/src/main/java/bio/terra/catalog/rawls/RawlsService.java:130]: I wonder if they are meant to be optional parameters, and the default is all?

**Review by s-rubenstein — APPROVED** [review]
LGTM

**Pull request merged**
